### PR TITLE
Relace moment to date-fns

### PR DESCRIPTION
**Pull request by otofune**
What
---
Use date-fns instead of moment, to use tree shaking. And remove many locales; include ja, en locales only.

Also this PR adds support to format comment's createdAt.

Screenshot from 2019-11-05 01-39-06

Why
---
As described in #473, we want to slim app.js size. (#535)
This PR reduces about 200KB size of it. Before this PR app.js had 2.28 MiB size, but now has only 2.05 MiB size in production build.

References
---
- date-fns format patterns: https://date-fns.org/v2.6.0/docs/format
- date-fns formatDistance words: https://date-fns.org/v2.6.0/docs/formatDistance

Tasks
---
- [x] Remove moment from server-side
- [x] Describe breaking changes that were happened by this changes
  + local format

### CODE REVIEW

File: client/components/Admin/Share/ShareList.tsx [hunk 1]
```diff
@@ -1,10 +1,10 @@
 import React from 'react'
 import { withTranslation, WithTranslation } from 'react-i18next'
-import moment from 'moment'
 import { Table, Alert } from 'reactstrap'
 import Pagination from 'components/Common/Pagination'
 import Crowi from 'client/util/Crowi'
 import { Share } from 'client/types/crowi'
+import { formatToLocaleString } from 'client/util/format-date'
```
Comment (lightnet328) [inline client/components/Admin/Share/ShareList.tsx]: In our modules, I think file name should be written in camel case.
(I feel this name was affected by `date-fns`.)

Comment (otofune) [reply client/components/Admin/Share/ShareList.tsx]: @lightnet328 
Thank you, I thought that only component files should be written in camel case. (eg: `frontend/crowi.ts`)
If really required, I will fix. :bowing_man:

File: client/util/format-date.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+function sanitizeDateInputForDateFns(input: DateInput) {
+  return typeof input === 'string' ? parseISO(input) : input
+}
+
+export default function format(date: DateInput, format: string, environment: Environment = defaultEnvironment) {
+  return dateFnsFormat(sanitizeDateInputForDateFns(date), format, { locale: environment.getLocale() })
+}
+
+export function formatToLocaleString(date: DateInput, environment: Environment = defaultEnvironment) {
```
Comment (otofune) [inline client/util/format-date.ts]: This is replacement of moment’s ‘llll’ format.
But date-fns’s ‘PPpp’ format doesn’t equals to moment’s one.

moment: Thu, Sep 4, 1986 8:30 PM
date-fns: May 29, 1453, 12:00:00 AM

It’s breaking change, but I think that it’s not critical problem.

File: client/util/format-date.test.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import { formatDistance, formatToLocaleString, defaultEnvironment, Environment } from './format-date'
+import { subDays } from 'date-fns'
+import { ja, enUS } from 'date-fns/locale'
```
Comment (otofune) [inline client/util/format-date.test.ts]: This reduces bundle size. But on the other hand, it drops many locales support.

I tried to resolve this problem by dynamic import with webpack. Crowi serves JS stuffs with /static path, however webpack’s dynamic loader loads splitted JS stuff with root path.

I think we can put off this problem. If we found correct webpack configuration in future, I will resolve this issue.

File: client/util/format-date.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/**
+ * OPINIONED date formatter to use browser locale by default
+ */
+
+import { format as dateFnsFormat, Locale, parseISO, formatDistance as dateFnsFormatDistance } from 'date-fns'
+import { enUS, ja } from 'date-fns/locale'
```
Comment (otofune) [inline client/util/format-date.ts]: This reduces bundle size. But on the other hand, it drops many locales support.

I tried to resolve this problem by dynamic import with webpack. Crowi serves JS stuffs with /static path, however webpack’s dynamic loader loads splitted JS stuff with root path.

I think we can put off this problem. If we found correct webpack configuration in future, I will resolve this issue.

**Comment by sotarok** [comment]
In `/admin/users`, error occurred:

```
index.js:372 Uncaught RangeError: Invalid time value
    at format (index.js:372)
    at UserTableRow (UserTableRow.tsx:58)
    at renderWithHooks (react-dom.development.js:12938)
    at mountIndeterminateComponent (react-dom.development.js:15020)
    at beginWork (react-dom.development.js:15625)
    at performUnitOfWork (react-dom.development.js:19312)
    at workLoop (react-dom.development.js:19352)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
format @ index.js:372
UserTableRow @ UserTableRow.tsx:58
renderWithHooks @ react-dom.development.js:12938
mountIndeterminateComponent @ react-dom.development.js:15020
beginWork @ react-dom.development.js:15625
performUnitOfWork @ react-dom.development.js:19312
workLoop @ react-dom.development.js:19352
callCallback @ react-dom.development.js:149
invokeGuardedCallbackDev @ react-dom.development.js:199
invokeGuardedCallback @ react-dom.development.js:256
replayUnitOfWork @ react-dom.development.js:18578
renderRoot @ react-dom.development.js:19468
performWorkOnRoot @ react-dom.development.js:20342
performWork @ react-dom.development.js:20254
performSyncWork @ react-dom.development.js:20228
requestWork @ react-dom.development.js:20097
scheduleWork @ react-dom.development.js:19911
dispatchAction @ react-dom.development.js:13599
_callee2$ @ UserPage.tsx:75
tryCatch @ runtime.js:45
invoke @ runtime.js:271
prototype.<computed> @ runtime.js:97
asyncGeneratorStep @ app.js:3968
_next @ app.js:3970
Promise.then (async)
asyncGeneratorStep @ app.js:3968
_next @ app.js:3970
(anonymous) @ app.js:3970
(anonymous) @ app.js:3970
fetchUsers @ UserPage.tsx:70
(anonymous) @ UserPage.tsx:86
commitHookEffectList @ react-dom.development.js:17283
commitPassiveHookEffects @ react-dom.development.js:17307
callCallback @ react-dom.development.js:149
invokeGuardedCallbackDev @ react-dom.development.js:199
invokeGuardedCallback @ react-dom.development.js:256
commitPassiveEffects @ react-dom.development.js:18774
wrapped @ scheduler-tracing.development.js:207
flushFirstCallback @ scheduler.development.js:107
flushWork @ scheduler.development.js:219
push../node_modules/scheduler/cjs/scheduler.development.js.channel.port1.onmessage @ scheduler.development.js:611

backend.js:6 The above error occurred in the <UserTableRow> component:
    in UserTableRow (created by UserTable)
    in tbody (created by UserTable)
    in table (created by UserTable)
    in UserTable (created by UserPage)
    in UserPage (created by Route)
    in Route (created by AdminPage)
    in Switch (created by AdminPage)
    in div (created by AdminPage)
    in div (created by AdminPage)
    in Router (created by BrowserRouter)
    in BrowserRouter (created by AdminPage)
    in AdminPage
```

File: client/components/Admin/Share/ShareList.tsx (hunk 1 above)
Comment (sotarok) [reply client/components/Admin/Share/ShareList.tsx]: Hmm, sorry it is not clearly written but it is good if `formatDate.ts` in turn with files in `lib/util` 🤔

**Comment by sotarok** [comment]
And also, it is not because of this PR but the comment may show server date?

I commented at 11pm but:

image

File: client/components/Admin/Share/ShareList.tsx (hunk 1 above)
Comment (otofune) [reply client/components/Admin/Share/ShareList.tsx]: :wrench: changed @ https://github.com/crowi/crowi/pull/594/commits/eafb051aab7f717395eb25089deb011714c903b4

**Comment by otofune** [comment]
I fixed above bug. But I will ask to review after I checked behavior about comments.

**Comment by otofune** [comment]
@sotarok 
I confirmed that raw creation date of comments are shown, also it is just as is.
I will fix it in this PR too.

**Comment by sotarok** [comment]
@otofune Sorry if it is caused by my environment but, it seems that `formatToLocaleString()` doesn't work correctly.

My env:

- image
    - (Crowi user setting)
- image
    - (stored value by i18next on localStorage)


And `formatToLocaleString` (formatted by `PPpp`) shows:

- image
- image

**Comment by otofune** [comment]
@sotraok At now, this formatter uses only browser’s locale. It’s same as previous behavior with moment. I thought it’s out of scope of this PR, so I created ticket #624.

**Comment by otofune** [comment]
@sotarok At now, this formatter uses only browser’s locale. It’s same as previous behavior with moment. I thought it’s out of scope of this PR, so I created ticket #624.

File: client/util/formatDate.test.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
+import { formatDistance, formatToLocaleString, defaultEnvironment, Environment } from './formatDate'
+import { subDays } from 'date-fns'
+import { ja, enUS } from 'date-fns/locale'
+
+// memo(otofune): This requires that process.env.TZ equals to 'UTC'. (;;)
```
Comment (lightnet328) [inline client/util/formatDate.test.ts]: Just a suggestion, how about overwriting `process.env.TZ` with this file to reduce the scope of impact?

File: package.json [hunk 6]
```diff
@@ -202,8 +202,8 @@
     "format": "eslint --fix '**/*.{ts,tsx,js,jsx}'",
     "lint": "eslint '**/*.{ts,tsx,js,jsx}'",
     "storybook": "start-storybook --ci -s public -p 6001",
-    "test": "DEBUG= jest --maxWorkers=5",
-    "test:client": "DEBUG= jest client --maxWorkers=5",
+    "test": "env -u DEBUG TZ=UTC jest --maxWorkers=5",
+    "test:client": "env -u DEBUG TZ=UTC jest client --maxWorkers=5",
     "test:server": "DEBUG= jest lib --maxWorkers=5",
```
Comment (lightnet328) [inline package.json]: Please use `env -u`👍

File: client/util/formatDate.test.ts (hunk 5 above)
Comment (otofune) [reply client/util/formatDate.test.ts]: After node v8, we can't change TZ with changing `process.env.TZ`. see also: https://github.com/nodejs/node/issues/3449

**Pull request merged**